lock/src/main/java/org/infinispan/concurrent/api/lock/impl/lock/ClusteredLockImpl.java

```diff
@@ -0,0 +1,203 @@
+                            ClusteredLockConfiguration configuration,
+                            AdvancedCache<ClusteredLockKey, ClusteredLockValue> clusteredLockCache,
+                            EmbeddedClusteredLockManager clusteredLockManager) {
+      this.name = name;
+      this.lockKey = lockKey;
+      this.lockConfiguration = configuration;
+      this.clusteredLockManager = clusteredLockManager;
+      this.pendingRequests = new ConcurrentLinkedQueue<>();
+      this.readWriteMap = ReadWriteMapImpl.create(FunctionalMapImpl.create(clusteredLockCache));
+      originator = clusteredLockCache.getCacheManager().getAddress();
+      clusteredLockCache.addListener(new LockReleasedListener(), new ClusteredLockFilter(lockKey));
+      clusteredLockCache.getCacheManager().addListener(new ClusterChangeListener());
+   }
+
+   public class LockRequestHolder<E> {
+      private final boolean tryLock;
+      private final Object requestor;
+      public final CompletableFuture<E> lockRequest;
+
+      public LockRequestHolder(Object owner, CompletableFuture<E> lockRequest) {
+         super();
+         this.requestor = owner;
+         this.lockRequest = lockRequest;
+         this.tryLock = false;
+      }
+
+      public LockRequestHolder(Object owner, CompletableFuture<E> lockRequest, boolean tryLock) {
+         super();
+         this.requestor = owner;
+         this.lockRequest = lockRequest;
+         this.tryLock = tryLock;
+      }
+
+      @Override
+      public String toString() {
+         return requestor.toString();
+      }
+
+      public boolean isDone() {
+         return lockRequest.isDone();
+      }
+
+      public void handleLockResult(Boolean lockResult, Throwable ex, boolean wait) {
+         if (ex != null) {
+            lockRequest.completeExceptionally(ex);
+            return;
+         }
+
+         if (lockResult == null) {
+            lockRequest.completeExceptionally(new InfinispanLockException("Lock result is null, something is wrong"));
+            return;
+         }
+
+         if (wait && lockResult) {
+            if (tryLock) {
+               lockRequest.complete((E) lockResult);
+            } else {
+               lockRequest.complete(null);
+            }
+         } else if (!wait) {
+            lockRequest.complete((E) lockResult);
+         }
+      }
+   }
+
+   @Listener(clustered = true)
+   class LockReleasedListener {
+
+      @CacheEntryModified
+      public void entryModified(CacheEntryModifiedEvent event) {
+         ClusteredLockValue value = (ClusteredLockValue) event.getValue();
+         if (value.getState() == ClusteredLockState.RELEASED) {
+            LockRequestHolder nextRequestor = null;
+            while (!pendingRequests.isEmpty() && (nextRequestor == null || nextRequestor.isDone()))
+               nextRequestor = pendingRequests.poll();
+
+            final LockRequestHolder requestor = nextRequestor;
+            clusteredLockManager.execute(() -> lock(requestor));
```

Comment:
executor can be remove once https://issues.jboss.org/browse/ISPN-8336 is fixed

Reply:
There is a race condition here with tryLock() with timer.
when a lock is released, the timer can be triggered and complete the future while the lock-request is in the network or in progress.
The lock is be blocked by this node in the cache and it can only be released from a unlock invocation of this node.